libnavigation-core/src/main/java/com/mapbox/navigation/core/telemetry/MapboxNavigationTelemetry.kt

```diff
@@ -116,42 +133,67 @@ internal object MapboxNavigationTelemetry {
-    private val dynamicFreeDriveValues = DynamicFreeDriveValues()
+
     private var locationEngineNameExternal: String = LocationEngine::javaClass.name
     private lateinit var locationsCollector: LocationsCollector
     private lateinit var sdkIdentifier: String
     private var logger: Logger? = null
     private val feedbackEventCacheMap = LinkedHashMap<String, NavigationFeedbackEvent>()
 
-    private var needHandleReroute = false
     private var sessionState: NavigationSession.State = IDLE
-    private var routeProgress: RouteProgress? = null
-    private var originalRoute: DirectionsRoute? = null
-    private var needStartSession = false
+
+    private val routeData = RouteData()
+
+    private class RouteData {
+        var routeProgress: RouteProgress? = null
+            set(value) {
+                field = value
+                onRouteDataChanged.invoke()
+            }
+        var originalRoute: DirectionsRoute? = null
+            set(value) {
+                field = value
+                onRouteDataChanged.invoke()
+            }
+
+        var needHandleDeparture = false
+            set(value) {
+                field = value
+                onRouteDataChanged.invoke()
+            }
+        var needHandleReroute = false
+
+        fun hasRouteAndRouteProgress(): Boolean {
+            return routeData.originalRoute != null && routeData.routeProgress != null
+        }
+    }
+
+    private var telemetryState: NavTelemetryState = NavTelemetryState.Stopped
+    private val isTelemetryRunning: Boolean
+        get() = telemetryState is NavTelemetryState.Running
 
     private val routesObserver = RoutesObserver { routes ->
         log("onRoutesChanged. size = ${routes.size}")
         routes.getOrNull(0)?.let {
             if (sessionState == ACTIVE_GUIDANCE) {
-                if (originalRoute != null) {
-                    if (needHandleReroute) {
-                        needHandleReroute = false
+                if (routeData.originalRoute != null) {
+                    if (routeData.needHandleReroute) {
+                        routeData.needHandleReroute = false
                         handleReroute(it)
                     } else {
                         log("handle ExternalRoute")
-                        sessionStop()
-                        originalRoute = it
-                        needStartSession = true
-                        startSessionIfNeedAndCan()
+                        handleCancelNavigation()
+                        resetLocalVariables()
+                        resetDynamicValues()
+                        routeData.originalRoute = it
+                        routeData.needHandleDeparture = true
                     }
                 } else {
-                    originalRoute = it
-                    needStartSession = true
-                    startSessionIfNeedAndCan()
+                    routeData.originalRoute = it
+                    routeData.needHandleDeparture = true
```

Comment:
nit: it feels unintuitive that this variable change initiatives the actual state check for departure. Same here and everywhere else.

Reply:
lets revisit it in the next iteration of refactoring of `MapboxNavigationTelemetry` https://github.com/mapbox/mapbox-navigation-android/issues/4744